### PR TITLE
ARROW-4853: [Rust] Array slice doesn't work on ListArray and StructArray

### PR DESCRIPTION
**Pull request by sunchao**
This fixes a few issues related to Array offset and slicing:

1. The invariant check in ListArray::from() is no longer valid for the case
  of offset. This removes the check.
2. Both `Array::is_valid` and `Array::is_null` should take into account of offset.
3. `StructArray::len()` should return `data.len()` instead of the length of the
  first child array. In the case of slicing, we'll not modify child array data so
  the previous approach will return wrong result.

### CODE REVIEW

**Comment by sunchao** [comment]
cc @nevi-me .

**Comment by nevi-me** [comment]
Thanks @sunchao , I'll review tonight (UTC).

**Review by andygrove — APPROVED** [review]
LGTM (once CI is green)

File: rust/arrow/src/array.rs
Comment (kszucs) [inline rust/arrow/src/array.rs]: @sunchao this fails the CI which indicates deeper issues, because `boxed_fields` serves only as a cache (at least in the cpp impl)

Comment (nevi-me) [reply rust/arrow/src/array.rs]: I was struggling to see what was causing struct-array tests to fail, thanks @kszucs

Comment (nevi-me) [reply rust/arrow/src/array.rs]: Hi @kszucs @sunchao, can a `StructArray` ever have a null value, or would it always be 0?

I found the issue here, when we create `StructArray` in `impl From<Vec<(Field, ArrayRef)>> for StructArray`, we check that all arrays have the same length, but never assign that length to the array buffer.

Setting a value for the length fixes the failure, but I'm not sure of what to do with null value. I'll check cpp codebase in the mean time.

```rust
[arrow\examples\dynamic_types.rs:62] &nested.data() = ArrayData {
    data_type: Struct(
        [
            Field {
                name: "a",
                data_type: Utf8,
                nullable: false
            },
            Field {
                name: "b",
                data_type: Float64,
                nullable: false
            },
            Field {
                name: "c",
                data_type: Float64,
                nullable: false
            }
        ]
    ),
    len: 5,
    null_count: 0,
    offset: 0,
    buffers: [],
    child_data: [
        ArrayData {
            data_type: Utf8,
            len: 5,
            null_count: 0,
            offset: 0,
            buffers: [
                Buffer {
                    data: BufferData {
                        ptr: 0x0000022cf1290e80,
                        len: 24
                    },
                    offset: 0
                },
                Buffer {
                    data: BufferData {
                        ptr: 0x0000022cf1290f00,
                        len: 5
                    },
                    offset: 0
                }
            ],
            child_data: [],
            null_bitmap: None
        },
        ArrayData {
            data_type: Float64,
            len: 5,
            null_count: 0,
            offset: 0,
            buffers: [
                Buffer {
                    data: BufferData {
                        ptr: 0x0000022cf1291840,
                        len: 40
                    },
                    offset: 0
                }
            ],
            child_data: [],
            null_bitmap: None
        },
        ArrayData {
            data_type: Float64,
            len: 5,
            null_count: 0,
            offset: 0,
            buffers: [
                Buffer {
                    data: BufferData {
                        ptr: 0x0000022cf1291940,
                        len: 40
                    },
                    offset: 0
                }
            ],
            child_data: [],
            null_bitmap: None
        }
    ],
    null_bitmap: None
}
```

Comment (nevi-me) [reply rust/arrow/src/array.rs]: https://github.com/apache/arrow/blob/master/cpp/src/arrow/array.cc#L373 indicates that this is possible, but I'm not getting joy from the docs (pyarrow and cpp)

**Comment by nevi-me** [comment]
@sunchao looks like in array.rs, the below works. It's nasty, and probably not efficient. I'll try something else tomorrow.

I have a test case of a struct array with nulls, I'll share the diff against your PR when I wake up in a few hours.

```rust
impl From<Vec<(Field, ArrayRef)>> for StructArray {
    fn from(v: Vec<(Field, ArrayRef)>) -> Self {
        let (field_types, field_values): (Vec<_>, Vec<_>) = v.into_iter().unzip();

        // Check the length of the child arrays
        let length = field_values[0].len();
        for i in 1..field_values.len() {
            assert_eq!(
                length,
                field_values[i].len(),
                "all child arrays of a StructArray must have the same length"
            );
        }

        // Manipulate bitmaps to find bits that are null for whole array elements
        // If we encounter a None in the Option<Bitmap>, it means that the struct can't be
        // null as a whole
        let bitmaps: Vec<Bitmap> = field_values
            .iter()
            .map(|array| array.data().null_bitmap().clone())
            .filter_map(|map| map)
            .collect();
        let nulls_and_bitmap: (usize, Option<Buffer>) = if bitmaps.len()
            != field_values.len()
        {
            (0, None)
        } else {
            // NOTE: The below only works if offset of each array in the struct is 0,
            // otherwise the problem's that bitmaps of different lengths are compared.

            // Any opinions on how to compare bitmaps that have differing offsets Chao?
            let mut first_bitmap: Bitmap = bitmaps[0].clone();
            for i in 1..field_values.len() {
                first_bitmap = first_bitmap.bitor(&bitmaps[i]).unwrap();
            }
            let valid_bits = first_bitmap.bits.data();
            // TODO if taking this approach, offsets have
            let null_count = length
                .checked_sub(bit_util::count_set_bits_offset(valid_bits, 0, length))
                .unwrap();

            (null_count, Some(first_bitmap.bits))
        };

        let builder = ArrayData::builder(DataType::Struct(field_types))
            .len(length)
            .null_count(nulls_and_bitmap.0)
            .child_data(field_values.into_iter().map(|a| a.data()).collect());
        let data = if let Some(buffer) = nulls_and_bitmap.1 {
            builder.null_bit_buffer(buffer).build()
        } else {
            builder.build()
        };
        Self::from(data)
    }
}
```

File: rust/arrow/src/array.rs
Comment (sunchao) [reply rust/arrow/src/array.rs]: @nevi-me yes StructArray can have null values - see https://arrow.apache.org/docs/format/Layout.html#struct-type.

**Comment by sunchao** [comment]
Thanks @nevi-me . For the case of `impl From<Vec<(Field, ArrayRef)>> for StructArray`, I'm not sure if we are going to handle null values though. It is important to differentiate between nulls in a struct array versus nulls in the child array of a struct array. I think we are going to handle the latter but the former cannot be done with this API. We should use `impl From<ArrayDataRef> for StructArray` instead.

**Comment by nevi-me** [comment]
Thanks @sunchao , perhaps we should document that we don't compute the null count and bitmask of struct arrays.

I wrote the below test, with:
* constructing a struct from 3 arrays
* the first 2 are slices
* the third is not sliced, but contains nulls

I found:
* the length of the array before slicing is now correct
* after slicing the struct array, its the 3 columns in the struct still report their length as 5 where it should be 3

```rust
#[test]
    fn test_struct_array_from_arrays_with_different_offsets() {
        let a: ArrayRef = Arc::new(Int32Array::from(vec![-1, 0, 1, 2, 3, 4, 5]));
        let sliced_a = a.slice(2, 5); // 1,2,3,4,5

        let b: ArrayRef = Arc::new(Float64Array::from(vec![
            -3.3, -2.2, -1.1, 0.0, 1.1, 2.2, 3.3, 4.4, 5.5,
        ]));
        let b_sliced = b.slice(3, 5);
        let nested = StructArray::from(vec![
            (Field::new("a", DataType::Int32, false), sliced_a),
            (Field::new("b", DataType::Float64, false), b_sliced),
            (
                Field::new("c", DataType::Float64, false),
                Arc::new(Float64Array::from(vec![
                    Some(1.1),
                    None,
                    None,
                    Some(4.4),
                    Some(5.5),
                ])),
            ),
        ]);

        // length of nested should be 5
        assert_eq!(5, nested.data().len());
        // null count based on 10100 | 10010 | 11010 = 11110
        assert_eq!(0, nested.data().null_count());
        let array = Arc::new(nested) as ArrayRef;
        let sliced = array.slice(2, 3);

        let sliced: &StructArray = array.as_any().downcast_ref::<StructArray>().unwrap();

        let a: &ArrayRef = sliced.column(0); // 3,4,5 (debugging this still shows 1,2,3,4,5) 
        let a: &Int32Array =
        a.as_any().downcast_ref::<Int32Array>().unwrap(); assert_eq!(3,
        a.value(0)); assert_eq!(4, a.value(1));
        assert_eq!(5, a.value(2));

        let b = sliced.column(1);
        let b = b.as_any().downcast_ref::<Float64Array>().unwrap(); // also 1.1, 2.2,
        ..., 5.5 assert_eq!(3, b.len());
        assert_eq!(3.3, b.value(0));
        assert_eq!(4.4, b.value(1));
        assert_eq!(5.5, b.value(2));

        let c = sliced.column(2);
        let c = c.as_any().downcast_ref::<Float64Array>().unwrap();
        dbg!(c);
        assert_eq!(1, c.null_count());
        assert_eq!(false, c.is_valid(0));
        assert_eq!(4.4, c.value(1));
        assert_eq!(5.5, c.value(2));
    }
```

**Comment by sunchao** [comment]
Good catch @nevi-me . We also need to consider offset when calling `StructArray::field`. Let me update the PR.

**Comment by nevi-me** [comment]
I'll work on https://issues.apache.org/jira/browse/ARROW-4886 once we merge this, then https://issues.apache.org/jira/browse/ARROW-4914 will go away

File: rust/arrow/src/array.rs
Comment (nevi-me) [inline rust/arrow/src/array.rs]: @sunchao are you happy that we open a JIRA to calculate null struct values in future as the `bitor` of the array's bitmasks?

File: rust/arrow/src/array.rs
Comment (nevi-me) [inline rust/arrow/src/array.rs]: as an observation, we are sometimes prone to adding a bit buffer without nulls, or the inverse when using `ArrayData::builder`. Should I open a JIRA for checking that if `null_count > 0` when calling build, the bit_buffer is also set?
I think this might catch subtle errors that we make in future.

File: rust/arrow/src/array.rs
Comment (nevi-me) [inline rust/arrow/src/array.rs]: I see that you reuse this bit_buffer on the boolean and int where they're not null. Should I disregard that, as the intention of the test is to show that the struct's offsets are considered correctly in the case of a slice?

File: rust/arrow/src/array.rs
Comment (nevi-me) [inline rust/arrow/src/array.rs]: nit: assuming that this evaluates to `10101`, do you mind changing it to something like `11101` or something where there isn't a repeating pattern? Checking the 3 ordered values of `101--` and `--101` would still result in a test passing without us knowing if it read the correct 3 bit masks.

File: rust/arrow/src/array.rs
Comment (nevi-me) [inline rust/arrow/src/array.rs]: Do you mind instead building the arrays with the `From<Vec<>>` builder, so they get their own butmaps? I think sharing the bitmap of the struct array shadows the intention that we should test the validity of the struct's column by its own bitmap instead of the struct's one.

**Review by nevi-me — COMMENTED** [review]
Thanks @sunchao, a few small questions and nits, otherwise this looks great!

File: rust/arrow/src/array.rs
Comment (sunchao) [reply rust/arrow/src/array.rs]: I think we need to decide whether an incoming vector such as:
```
[
  (BooleanType, [false, null, true, null, false]),
  (Int32Type, [32, null, null, null, 28])
]
```
Should be converted to:
```
(StructType, [(false, 32), null, (true, null), null, (false, 28)])
```

or:
```
(StructType, [(false, 32), (null, null), (true, null), (null, null), (false, 28)])
```

These two may have subtle differences. Feel free to open a JIRA for this and maybe we can have more discussion there.

File: rust/arrow/src/array.rs
Comment (sunchao) [reply rust/arrow/src/array.rs]: Good suggestion. Will change.

File: rust/arrow/src/array.rs
Comment (sunchao) [reply rust/arrow/src/array.rs]: Sure. I'll use different bit buffer for boolean and int.

File: rust/arrow/src/array.rs
Comment (sunchao) [reply rust/arrow/src/array.rs]: Can you elaborate on this? If `null_count` is not specified in the input, it is calculated from `bit_buffer` so they should be consistent. Otherwise we'll just use `null_count`.

File: rust/arrow/src/array.rs
Comment (sunchao) [reply rust/arrow/src/array.rs]: We can pass different bitmaps for different columns even with the current approach. Let me update the PR.

File: rust/arrow/src/array.rs
Comment (nevi-me) [reply rust/arrow/src/array.rs]: Great point, I didn't inspect the function closely, my apologies for that. 
I was trying to say that nothing stops the user from doing the below:

```rust
ArrayData::builder(dt)
  .len(n)
  .null_count(Some(3)) // only setting null count, but not supplying bitmap
  .buffers(vec![my_buffer])
  .build();
```

The inverse where I supply just the bitmap isn't an issue because the null_count is calculated like you mention. It's the above case where I'm unsure of what should happen.

File: rust/arrow/src/array.rs
Comment (nevi-me) [reply rust/arrow/src/array.rs]: Thanks, I'll open a JIRA. I'm curious to know how C++ does things, I'll maybe also ask in the mailing list post 0.13 release

Comment (sunchao) [reply rust/arrow/src/array.rs]: I think this is unique in Rust - C++ doesn't have a API to convert a vector of `(Field, ArrayRef)` into a `StructArray`.

File: rust/arrow/src/array.rs
Comment (sunchao) [reply rust/arrow/src/array.rs]: Yes in this case it won't verify whether the bit buffer is consistent with null count. It will use the value from null count for `null_count()` but bit buffer for `is_valid()` and `is_null()`. 

With that said, I think the purpose of the null count parameter is to avoid calculating it from the bit buffer. To do the extra step of validating it seems to defeat that purpose.

Comment (nevi-me) [reply rust/arrow/src/array.rs]: Okay, I understand. I've seen that in cpp and JS the null count is often left as unknown (-1).

**Review by nevi-me — APPROVED** [review]
Thanks @sunchao

**Pull request closed without merging**
